File: src/creational/abstract_factory/base/MazeFactory.java
package creational.abstract_factory.base;

public class MazeFactory {
    public Maze makeMaze() {
        return new Maze();
    }
    public Wall makeWall() {
        return new Wall();
    }
    public Room makeRoom(int n) {
        return new Room(n);
    }
    public Door makeDoor(Room r1, Room r2) {
        return new Door(r1, r2);
    }
}

File: src/creational/abstract_factory_kotlin/MazeFactory.kt
package creational.abstract_factory_kotlin

import base.maze.Door
import base.maze.Maze
import base.maze.Room
import base.maze.Wall

/**
 * Different implementations will extend this MazeFactory class.
 *
 * By overriding the "make" methods and using its own custom Walls, Rooms and Doors.
 *
 * Pass each MazeFactory instance to the MazeGame.createMaze() method.
 */
open class MazeFactory {
    fun makeMaze() = Maze()
    open fun makeWall() = Wall()
    open fun makeRoom(n: Int) = Room(n)
    open fun makeDoor(r1: Room, r2: Room) = Door(r1, r2)
}
